src/watchman_client.js

```diff
@@ -0,0 +1,492 @@
+'use strict';
+
+var watchman = require('fb-watchman');
+
+/**
+ * Constants
+ */
+
+/**
+ * Singleton that provides a public API for a connection to a watchman instance for 'sane'.
+ * It tries to abstract/remove as much of the boilerplate processing as necessary
+ * from WatchmanWatchers that use it. In particular, they have no idea whether
+ * we're using 'watch-project' or 'watch', what the 'project root' is when
+ * we internally use watch-project, whether a connection has been lost
+ * and reestablished, etc. Also switched to doing things with promises and known-name
+ * methods in WatchmanWatcher, so as much information as possible can be kept in 
+ * the WatchmanClient, ultimately making this the only object listening directly
+ * to watchman.Client, then forwarding appropriately (via the known-name methods) to 
+ * the relevant WatchmanWatcher(s).
+ * 
+ * Note: WatchmanWatcher also added a 'watchmanPath' option for use with the sane CLI.
+ * Because of that, we actually need a map of watchman binary path to WatchmanClient instance.
+ * That is set up in getInstance(). Once the WatchmanWatcher has a given client, it doesn't
+ * change.
+ *
+ * @class WatchmanClient
+ * @param String watchmanBinaryPath 
+ * @public
+ */
+
+function WatchmanClient(watchmanBinaryPath) {
+
+  // define/clear some local state. The properties will be initialized
+  // in _handleClientAndCheck(). This is also called again in _onEnd when
+  // trying to reestablish connection to watchman.
+  this._clearLocalVars(); 
+
+  this._watchmanBinaryPath = watchmanBinaryPath;
+  this._clientListeners = null;  // direct listeners from here to watchman.Client.
+}
+
+// Define 'wildmatch' property, which must be available when we call the
+// WatchmanWatcher.createOptions() method. 
+Object.defineProperty(WatchmanClient.prototype, 'wildmatch', {
+  get() { return this._wildmatch; }
+});
+
+/**
+ * Called from WatchmanWatcher (or WatchmanClient during reconnect) to create
+ * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+ * watchman.Client, to be handled here.
+ */
+WatchmanClient.prototype.subscribe = function(watchmanWatcher, root) {
+  let subscription;
+  let watcherInfo;
+  
+  return this._setupClient()
+    .then(() => {
+      watcherInfo = this._createWatcherInfo(watchmanWatcher);
+      subscription = watcherInfo.subscription;
+      return this._watch(subscription, root);
+    })
+    .then((resp) => this._clock(subscription))
+    .then((data) => this._subscribe(subscription))
+    .catch((error) => {
+      console.error("Failed creating subscription for root " + root);
```

Comment:
should this log be here, or should the consumer handle logging? As implemented, the user cannot avoid this log, even if they handle the error correctly.

Reply:
What do you suggest on this one? Generating an 'error' event to the watcher? Not trapping the 'catch' at all, so the original error just goes back as a rejected promise? (It will anyway).